File: baekjoon/7000/7585.cpp
#include<bits/stdc++.h>
using namespace std;
#define loop(i,n) for(int i=0;i<n;++i)

bool legal(string& s)
{
    stack<char> stk;
    int len = s.size();
    loop(i, len) {
        char c = s[i];
        if (c == '(' || c == '{' || c == '[') stk.push(c);
        else if (c == ')') {
            if (stk.empty() || stk.top() != '(') return false;
            stk.pop();
        }
        else if (c == '}') {
            if (stk.empty() || stk.top() != '{') return false;
            stk.pop();
        }
        else if (c == ']') {
            if (stk.empty() || stk.top() != '[') return false;
            stk.pop();
        }
    }
    if (!stk.empty()) return false;
    return true;
}
int main()
{
    cin.tie(NULL), cout.tie(NULL);
    ios::sync_with_stdio(false);

    string s;
    while (1) {
        getline(cin, s);
        if (s == "#") break;
        if (legal(s)) cout << "Legal\n";
        else cout << "Illegal\n";
    }
    return 0;
}

File: baekjoon/7000/7585.rs
use std::io::*;

fn main() {
    let mut bw = BufWriter::new(stdout().lock());
    let mut s = String::new();
    stdin().read_to_string(&mut s).ok();
    let mut it = s.split('\n');
    let mut stack = vec![];
    loop {
        let s = it.next().unwrap();
        if s == "#" {
            break;
        }
        stack.clear();
        let ok = s.bytes().all(|x| match x {
            b'(' | b'{' | b'[' => {
                stack.push(x);
                true
            }
            b')' => stack.pop().is_some_and(|y| y == b'('),
            b'}' => stack.pop().is_some_and(|y| y == b'{'),
            b']' => stack.pop().is_some_and(|y| y == b'['),
            _ => true,
        }) && stack.is_empty();
        writeln!(bw, "{}", if ok { "Legal" } else { "Illegal" }).ok();
    }
}
